### PR TITLE
Implemented UrlFilterApplier component

### LINKED ISSUE #1501: Add ability to set the filters on UI grid via GET parameters of request
TBD

### PR DESCRIPTION
**Pull request by gabrieldagama**
### Description (*)
Implemented a new component to apply filters on a grid via GET URL parameter. Added the component to the product, cms_page, and cms_block grids.

### Fixed Issues (if relevant)

1. Fixes magento/adobe-stock-integration#1501

### Manual testing scenarios (*)

1. Open one of the grids mentioned above
2. Add the filters parameter to the URL, example for product: ?filters[name]=product123
3. The grid should be loaded with a name active filter with value = product123

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Test/Mftf/Test/AdminProductGridUrlFilterApplierTest.xml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+<tests xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/testSchema.xsd">
+    <test name="AdminProductGridUrlFilterApplierTest">
+        <annotations>
+            <features value="Catalog"/>
+            <stories value="Filter product using GET URL parameter"/>
+            <title value="Verify that filter is applied on product grid when filters parameter is set on url"/>
+            <description value="Accessing product grid url with filters parameter"/>
+            <severity value="MAJOR"/>
+            <testCaseId value="to-be-added"/>
```
Comment (sivaschenko) [inline app/code/Magento/Catalog/Test/Mftf/Test/AdminProductGridUrlFilterApplierTest.xml]: Test case id needs to be added

File: app/code/Magento/Catalog/Test/Mftf/Test/AdminProductGridUrlFilterApplierTest.xml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        <after>
+            <deleteData createDataKey="createSimpleProduct" stepKey="deleteProduct"/>
+            <actionGroup ref="AdminLogoutActionGroup" stepKey="logoutOfAdmin"/>
+        </after>
+        <amOnPage url="{{AdminProductIndexPage.url}}?filters[name]=$$createSimpleProduct.name$$" stepKey="navigateToProductGridWithFilters"/>
+        <waitForPageLoad stepKey="waitForProductGrid"/>
+        <see selector="{{AdminProductGridSection.productGridNameProduct($$createSimpleProduct.name$$)}}" userInput="$$createSimpleProduct.name$$" stepKey="seeProduct"/>
+        <seeElement selector="{{AdminProductGridFilterSection.enabledFilters}}" stepKey="seeEnabledFilters"/>
+        <see selector="{{AdminProductGridFilterSection.enabledFilters}}" userInput="Name: $$createSimpleProduct.name$$" stepKey="seeProductNameFilter"/>
+        <actionGroup ref="ClearFiltersAdminDataGridActionGroup" stepKey="clearGridFilter"/>
```
Comment (sivaschenko) [inline app/code/Magento/Catalog/Test/Mftf/Test/AdminProductGridUrlFilterApplierTest.xml]: I'd move this action group to `after` section

File: app/code/Magento/Cms/Test/Mftf/Test/AdminCmsBlockGridUrlFilterApplierTest.xml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+<tests xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/testSchema.xsd">
+    <test name="AdminCmsBlockGridUrlFilterApplierTest">
+        <annotations>
+            <features value="Cms"/>
+            <stories value="Filter CMS block using GET URL parameter"/>
+            <title value="Verify that filter is applied on block grid when filters parameter is set on url"/>
+            <description value="Accessing block grid url with filters parameter"/>
+            <severity value="MAJOR"/>
+            <testCaseId value="to-be-added"/>
```
Comment (sivaschenko) [inline app/code/Magento/Cms/Test/Mftf/Test/AdminCmsBlockGridUrlFilterApplierTest.xml]: Test case need to be added

File: app/code/Magento/Cms/Test/Mftf/Test/AdminCmsPageGridUrlFilterApplierTest.xml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        <after>
+            <deleteData createDataKey="createPage" stepKey="deletePage"/>
+            <actionGroup ref="AdminLogoutActionGroup" stepKey="logoutOfAdmin"/>
+        </after>
+        <amOnPage url="{{CmsPagesPage.url}}?filters[title]=$$createPage.title$$" stepKey="navigateToPageGridWithFilters"/>
+        <waitForPageLoad stepKey="waitForPageGrid"/>
+        <see selector="{{CmsPagesPageActionsSection.pagesGridRowByTitle($$createPage.title$$)}}" userInput="$$createPage.title$$" stepKey="seePage"/>
+        <seeElement selector="{{CmsPagesPageActionsSection.activeFilter}}" stepKey="seeEnabledFilters"/>
+        <see selector="{{CmsPagesPageActionsSection.activeFilter}}" userInput="Title: $$createPage.title$$" stepKey="seePageTitleFilter"/>
+        <actionGroup ref="ClearFiltersAdminDataGridActionGroup" stepKey="clearGridFilter"/>
```
Comment (sivaschenko) [inline app/code/Magento/Cms/Test/Mftf/Test/AdminCmsPageGridUrlFilterApplierTest.xml]: I'd move this action group to `after` section

File: app/code/Magento/Ui/view/base/web/js/grid/url-filter-applier.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+define([
+    'uiComponent',
+    'underscore'
+], function (Component, _) {
+    'use strict';
+
+    return Component.extend({
+        defaults: {
+            filterProvider: 'componentType = filters',
```
Comment (sivaschenko) [inline app/code/Magento/Ui/view/base/web/js/grid/url-filter-applier.js]: how this component is behaving on a page with multiple grids?
I.e. media gallery + adobe stock

File: app/code/Magento/Ui/view/base/web/js/grid/url-filter-applier.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+            if (_.isUndefined(this.filterComponent())) {
+                setTimeout(function () {
+                    this.apply();
+                }.bind(this), 100);
+
+                return;
+            }
+            if (Object.keys(urlFilter).length) {
+                this.filterComponent().setData(urlFilter, false);
```
Comment (sivaschenko) [inline app/code/Magento/Ui/view/base/web/js/grid/url-filter-applier.js:50]: How this component is behaving if the column/filter in the URL is not present or hidden on the grid?
Or in case there are URL parameters that are not related to the filters?

File: app/code/Magento/Cms/Test/Mftf/Test/AdminCmsPageGridUrlFilterApplierTest.xml [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+<tests xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/testSchema.xsd">
+    <test name="AdminCmsPageGridUrlFilterApplierTest">
+        <annotations>
+            <features value="CmsPage"/>
+            <stories value="Filter CMS page using GET URL parameter"/>
+            <title value="Verify that filter is applied on page grid when filters parameter is set on url"/>
+            <description value="Accessing page grid url with filters parameter"/>
+            <severity value="MAJOR"/>
+            <testCaseId value="to-be-added"/>
```
Comment (sivaschenko) [inline app/code/Magento/Cms/Test/Mftf/Test/AdminCmsPageGridUrlFilterApplierTest.xml]: Test case needs to be added

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Great contribution @gabrieldagama ! Thanks for the double test coverage! Please take a look at my comments

File: app/code/Magento/Cms/Test/Mftf/Test/AdminCmsBlockGridUrlFilterApplierTest.xml (hunk 3 above)
Comment (gabrieldagama) [reply app/code/Magento/Cms/Test/Mftf/Test/AdminCmsBlockGridUrlFilterApplierTest.xml]: Yes, @chalov-anton can you provide the test cases for this issue?

File: app/code/Magento/Ui/view/base/web/js/grid/url-filter-applier.js (hunk 5 above)
Comment (gabrieldagama) [reply app/code/Magento/Ui/view/base/web/js/grid/url-filter-applier.js]: Good point, will have a look on this, probably will have to introduce a parameter to handle the component namespace.

File: app/code/Magento/Ui/view/base/web/js/grid/url-filter-applier.js (hunk 6 above)
Comment (gabrieldagama) [reply app/code/Magento/Ui/view/base/web/js/grid/url-filter-applier.js:50]: - If the column/filter is not present it should behave the same way as the normal filter work, but I will test this scenario.
- If other parameters are added these are ignored. We are just looking for the `filters` parameter.

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-7778](https://jira.corp.magento.com/browse/ENGCOM-7778) has been created to process this Pull Request
:eight_spoked_asterisk: @sivaschenko, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed

Manual scenario on Cucumber Studio - https://studio.cucumber.io/projects/131313/test-plan/folders/1320712/scenarios/4931106

**Comment by engcom-Charlie** [comment]
@magento run all tests

File: app/code/Magento/Cms/view/adminhtml/layout/cms_block_index.xml [hunk 8]
```diff
@@ -9,6 +9,7 @@
     <body>
         <referenceContainer name="content">
             <uiComponent name="cms_block_listing"/>
+            <block class="Magento\Backend\Block\Template" template="Magento_Cms::url_filter_applier.phtml" />
```
Comment (konarshankar07) [inline app/code/Magento/Cms/view/adminhtml/layout/cms_block_index.xml]: I think we can add `Urlfilter` component in ui_component instead of creating a new block. Please check the similar implementation
https://github.com/magento/adobe-stock-integration/blob/f8f7257d30bf54646536491ad5366ae082c13570/AdobeStockImageAdminUi/view/adminhtml/ui_component/adobe_stock_images_listing.xml#L55-L70
Let me know your thoughts
Thanks

Comment (sivaschenko) [reply app/code/Magento/Cms/view/adminhtml/layout/cms_block_index.xml]: @konarshankar07 I think both approaches are fine

Comment (konarshankar07) [reply app/code/Magento/Cms/view/adminhtml/layout/cms_block_index.xml]: @sivaschenko ... Yes, you are right but my concerns is the namespace are hard coded in the `url_filter_applier.phtml` file which can be easily exported using ui_component.
